### ISSUE TITLE
Improve parsing of numbers in util/str.rs

### ISSUE THREAD

**Issue by KiChjang** [servo/servo#10861]
#10608 added support to parse double values from DOMStrings, but the mechanism it uses is messy - we have to keep track of how many digits we've parsed for the units part and the fraction part. We should make use of `take_while_ref` from the itertools crate.

Code: `components/util/str.rs`

**Comment by SunyDays** [comment]
I'd like to work on it.
I understand correctly, that tracking number of digits in unit and fraction parts need add to functions read_fraction and read_exponent for control that value which we parse can be stored in f64?

**Comment by KiChjang** [comment]
No, that's not the issue here. We want to _not_ keep track of the number of digits parsed in `read_fraction` or `read_numbers`; instead, it should simply mutate the iterator, fast-forwarding to the index after the last digit was parsed. `take_while_ref` from the `itertools` crate is perfect for this, but it looks like there are some trait bounds that are not satisfied when I first tried to implement it in #10608.

**Comment by KiChjang** [comment]
Previous attempt at #10974.
